src/Microsoft.VisualStudio.ProjectSystem.Managed/Threading/Tasks/CancellationSeries.cs

```diff
@@ -0,0 +1,112 @@
+        public CancellationSeries(CancellationToken token = default)
+        {
+            _superToken = token == default 
+                ? null 
+                : new[] { token };
+        }
+
+        /// <summary>
+        /// Gets the next <see cref="CancellationToken"/> in the series, ensuring the last issued
+        /// token (if any) is cancelled before this method returns.
+        /// </summary>
+        /// <param name="token">An optional cancellation token that, when cancelled, cancels the
+        /// returned token.</param>
+        /// <returns>
+        /// A cancellation token that will be cancelled when either:
+        /// <list type="bullet">
+        /// <item><see cref="GetToken"/> is called again</item>
+        /// <item>The token passed to this method (if any) is cancelled</item>
+        /// <item>The token passed to the constructor (if any) is cancelled</item>
+        /// <item><see cref="Dispose"/> is called</item>
+        /// </list>
+        /// </returns>
+        /// <exception cref="ObjectDisposedException">This object has been disposed.</exception>
+        public CancellationToken GetToken(CancellationToken token = default)
+        {
+            // Create the next CTS. There are four scenarios, depending upon whether
+            // a super token was given to the constructor, and whether a token was
+            // passed to this method.
+
+            CancellationTokenSource nextSource = _superToken == null
+                ? token == default
+                    ? new CancellationTokenSource()
+                    : CancellationTokenSource.CreateLinkedTokenSource(token)
+                : token == default
+                    ? CancellationTokenSource.CreateLinkedTokenSource(_superToken)
+                    : CancellationTokenSource.CreateLinkedTokenSource(token, _superToken[0]);
+
+            CancellationToken nextToken = nextSource.Token;
+            CancellationTokenSource priorSource = Volatile.Read(ref _cts);
+
+            // Loop until we succeed in swapping a known prior for our next.
+            // This ensures a proper sequence is observed, which is important
+            // as otherwise we may cancel a prior twice, which would coincide
+            // with a prior not being cancelled at all. We need a strict order
+            // to maintain one-in-one-out.
+            while (true)
+            {
+                if (priorSource == null)
+                {
+                    nextSource.Dispose();
+
+                    throw new ObjectDisposedException(nameof(CancellationSeries));
+                }
+
+                CancellationTokenSource snapshot = Interlocked.CompareExchange(ref _cts, nextSource, priorSource);
+
+                if (ReferenceEquals(snapshot, priorSource))
+                {
+                    break;
+                }
+
+                priorSource = snapshot;
+            }
+
+            priorSource.Cancel();
```

Comment:
Maybe this isn't possible based on construction, but do you need to handle the case where a delegate registered with a cancellation token throws an unhandled exception?  That would emerge from here and prevent priorSource from being disposed and prevent the token from being returned.  Does that matter?

Reply:
This sounds entirely reasonable. I'll look into the mechanics.

For now the only protection that comes to mind is catching exceptions thrown from `Cancel`, disposing both sources and rethrowing.

Reply:
Fixed in 6d9b69a7f82774095a81960b2ad4dfe80c89bb24, along with a test.

I couldn't find a way to verify that `priorSource` was disposed via public APIs on `CancellationToken`.